### ISSUE TITLE
Error when compiling with declaration flag on because of alias inlining

### ISSUE THREAD

**Issue by gcnew** [Microsoft/TypeScript#15052]
**TypeScript Version:**  nightly (2.3.0-dev.20170406)

**Code**

_tsconfig.json_
```ts
{
    "compilerOptions": {
        "noUnusedLocals": true
    },

    "include": [
        "*.ts"
    ]
}

```

_A.ts_
```ts

export { Either, Fail, Magic, Inliner, InlinedMagic }

type Either<L, R> = { tag: 'left',  value: L }
                  | { tag: 'right', value: R }

interface Fail {
    message: string;
}

type Magic<T> = (x: T) => Either<Fail, T>

type Inliner<T> = T

type InlinedMagic<T> = Inliner<Magic<T>>
```

_B.ts_
```ts
import {
    // if `Fail` is uncommented, the compiler complains because of `noUnusedLocals`
    InlinedMagic /*, Fail */ 
} from './A'

export { repro }

function repro() {
//       ~~~~~
// B.ts(9,10): error TS4058: Return type of exported function has or is using name 'Fail'
//             from external module "/projects/repro/A" but cannot be named.

    return <T>(): InlinedMagic<T> => x => ({ tag: 'right', value: x });
}
```

**Prerequisites for reproduction:**
 - `Fail` must be an interface, if it is an alias it gets inlined and no error is raised
 - the return type of the inner function must be inlined for the issue to surface - double alias indirection triggers inlining
 - the return type of the outer function must be inferred. User annotations are used as is in the declaration file otherwise.

**Expected behavior:**
Either `noUnusedLocals` should recognise that `Fail` is used because of inlining, or the type inferencer shouldn't inline the type of the returned inner function.

**Actual behavior:**
```
B.ts(9,10): error TS4058: Return type of exported function has or is using name 'Fail' from external module "/projects/repro/A" but cannot be named.
```

**Comment by mhegazy** [comment]
this is a duplicate of https://github.com/Microsoft/TypeScript/issues/9944.

you need to 1. import `Fail` and 2. use it in the type annotation of `repro`

**Comment by gcnew** [comment]
Yes, I figured the solution out and then went on with reducing the problem into a repro. Although the linked issue is relevant and will fix this one as well, I think the most obvious and correct resolution in this specific case to use the `InlineMagic` alias as is, rather than inlining it.
